### PR TITLE
Add ConvNeXT

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [ConvNeXT](https://github.com/facebookresearch/ConvNeXt) to the library, a convnet inspired by Transformers.

To do:

- [x] rename nielsr to facebook to fix tests
- [x] remove inference.py script
- [x] remove PushToHubMixin hack from feature extractor
- [x] fix "gamma" in `from_pretrained` method. Right now, if a model includes a parameter named "gamma", they are not initialized from the hub, due to [this line](https://github.com/huggingface/transformers/blob/833635e25997a00117076c50bde5c45f9b883ada/src/transformers/modeling_utils.py#L1490).
- [x] discuss model output classes. It feels a bit weird to talk about "hidden states" for convolutional models, most people use the term "features" or "feature maps". These are of shape (batch_size, num_channels, a certain height, a certain width) at each stage, rather than (batch_size, seq_len, hidden_size) at each layer of a Transformer. For now, they are still called `hidden_states` in this PR.

### CODE REVIEW

File: docs/source/model_doc/convnext.mdx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+(e.g., Swin Transformers) that reintroduced several ConvNet priors, making Transformers practically viable as a generic vision backbone and demonstrating remarkable performance on a wide
+variety of vision tasks. However, the effectiveness of such hybrid approaches is still largely credited to the intrinsic superiority of Transformers, rather than the inherent inductive
+biases of convolutions. In this work, we reexamine the design spaces and test the limits of what a pure ConvNet can achieve. We gradually "modernize" a standard ResNet toward the design
+of a vision Transformer, and discover several key components that contribute to the performance difference along the way. The outcome of this exploration is a family of pure ConvNet models
+dubbed ConvNeXt. Constructed entirely from standard ConvNet modules, ConvNeXts compete favorably with Transformers in terms of accuracy and scalability, achieving 87.8% ImageNet top-1 accuracy
+and outperforming Swin Transformers on COCO detection and ADE20K segmentation, while maintaining the simplicity and efficiency of standard ConvNets.*
+
+Tips:
+
+- One can use the ['AutoFeatureExtractor`] API to prepare images for the model.
```
Comment (sgugger) [inline docs/source/model_doc/convnext.mdx:32]:
```suggestion
- One can use the [`AutoFeatureExtractor`] API to prepare images for the model.
```

File: docs/source/model_doc/convnext.mdx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+This model was contributed by [nielsr](<https://huggingface.co/nielsr). The original code can be found [here](https://github.com/facebookresearch/ConvNeXt).
```
Comment (sgugger) [inline docs/source/model_doc/convnext.mdx:34]:
```suggestion
This model was contributed by [nielsr](https://huggingface.co/nielsr). The original code can be found [here](https://github.com/facebookresearch/ConvNeXt).
```

File: src/transformers/__init__.py [hunk 3]
```diff
@@ -633,6 +634,7 @@
     _import_structure["modeling_utils"] = ["Conv1D", "PreTrainedModel", "apply_chunking_to_forward", "prune_layer"]
 
     # PyTorch models structure
+
```
Comment (sgugger) [inline src/transformers/__init__.py:637]:
```suggestion
```

File: src/transformers/__init__.py [hunk 4]
```diff
@@ -2659,6 +2671,7 @@
         from .utils.dummy_pytorch_quantization_and_torch_objects import *
 
     if is_torch_available():
+
```
Comment (sgugger) [inline src/transformers/__init__.py:2674]:
```suggestion
```

File: src/transformers/__init__.py [hunk 5]
```diff
@@ -832,6 +834,15 @@
             "load_tf_weights_in_convbert",
         ]
     )
+    _import_structure["models.convnext"].extend(
+        [
+            "CONVNEXT_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "ConvNextBlock",
```
Comment (sgugger) [inline src/transformers/__init__.py:840]: Not sure this one needs to be public, but it's your call.

File: src/transformers/models/convnext/configuration_convnext.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+CONVNEXT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "facebook/convnext-tiny-224": "https://huggingface.co/facebook/convnext-tiny-224/resolve/main/config.json",
+    # See all ConvNeXT models at https://huggingface.co/models?filter=convnext
+}
+
+
+class ConvNextConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`~ConvNextModel`]. It is used to instantiate an
```
Comment (sgugger) [inline src/transformers/models/convnext/configuration_convnext.py:31]:
```suggestion
    This is the configuration class to store the configuration of a [`ConvNextModel`]. It is used to instantiate an
```

File: src/transformers/models/convnext/configuration_convnext.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    ConvNeXT model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the ConvNeXT
+    [facebook/convnext-tiny-224](https://huggingface.co/facebook/convnext-tiny-224) architecture.
+
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+    Args:
+        num_channels (`int`, *optional*, defaults to `3`):
```
Comment (sgugger) [inline src/transformers/models/convnext/configuration_convnext.py:40]:
```suggestion
        num_channels (`int`, *optional*, defaults to 3):
```
No code for numbers.

File: src/transformers/models/convnext/inference.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import torch
```
Comment (sgugger) [inline src/transformers/models/convnext/inference.py:1]: This script should not be added here.

File: src/transformers/models/convnext/modeling_convnext.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+        return x
+    keep_prob = 1 - drop_prob
+    shape = (x.shape[0],) + (1,) * (x.ndim - 1)  # work with diff dim tensors, not just 2D ConvNets
+    random_tensor = keep_prob + torch.rand(shape, dtype=x.dtype, device=x.device)
+    random_tensor.floor_()  # binarize
+    output = x.div(keep_prob) * random_tensor
+    return output
+
+
+class DropPath(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:89]: Should be named `ConvNextDropPath`

File: src/transformers/models/convnext/modeling_convnext.py [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+            x = self.weight[:, None, None] * x + self.bias[:, None, None]
+            return x
+
+
+class ConvNextBlock(nn.Module):
+    """This corresponds to the ConvNextBlock class in the original implementation.
+
+    There are two equivalent implementations: (1) DwConv -> LayerNorm (channels_first) -> 1x1 Conv -> GELU -> 1x1 Conv;
+    all in (N, C, H, W) (2) DwConv -> Permute to (N, H, W, C); LayerNorm (channels_last) -> Linear -> GELU -> Linear;
+    Permute back
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:132]: Would be more readable put it in a list.

File: src/transformers/models/convnext/modeling_convnext.py [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(CONVNEXT_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=ConvNextModelOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        pixel_values=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:275]: Fits in one line.

File: src/transformers/models/convnext/modeling_convnext.py [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+        ```"""
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if pixel_values is None:
+            raise ValueError("You have to specify pixel_values")
+
+        print("Shape of pixel values:", pixel_values.shape)
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:304]: Should be removed.

File: src/transformers/models/convnext/modeling_convnext.py [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(CONVNEXT_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=SequenceClassifierOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        pixel_values=None,
+        labels=None,
+        output_hidden_states=None,
+        return_dict=None,
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:355]: Fits in one line.

File: src/transformers/models/convnext/modeling_convnext.py [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+        >>> # model predicts one of the 1000 ImageNet classes
+        >>> predicted_class_idx = logits.argmax(-1).item()
+        >>> print("Predicted class:", model.config.id2label[predicted_class_idx])
+        ```"""
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        outputs = self.convnext(
+            pixel_values,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:390]: Fits in one line.

File: src/transformers/models/convnext/modeling_convnext.py [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+
+        loss = None
+        if labels is not None:
+            if self.num_labels == 1:
+                #  We are doing regression
+                loss_fct = MSELoss()
+                loss = loss_fct(logits.view(-1), labels.view(-1))
+            else:
+                loss_fct = CrossEntropyLoss()
+                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```
Comment (sgugger) [inline src/transformers/models/convnext/modeling_convnext.py:405]: This is not using the `problem_type` config argument, and it should.

File: tests/test_modeling_convnext.py [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+        result = model(pixel_values, labels=labels)
+        self.parent.assertEqual(result.logits.shape, (self.batch_size, self.type_sequence_label_size))
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
Comment (sgugger) [inline tests/test_modeling_convnext.py:123]: Fits in one line.

**Review by sgugger — APPROVED** [review]
Great new addition!

File: src/transformers/__init__.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/__init__.py:637]: The CookieCutter seems to always insert this line. Can we fix this [here](https://github.com/huggingface/transformers/blob/master/templates/adding_a_new_model/cookiecutter-template-%7B%7Bcookiecutter.modelname%7D%7D/to_replace_%7B%7Bcookiecutter.lowercase_modelname%7D%7D.py)?

File: src/transformers/__init__.py (hunk 4 above)
Comment (NielsRogge) [reply src/transformers/__init__.py:2674]: Same comment here.

File: src/transformers/__init__.py (hunk 5 above)
Comment (NielsRogge) [reply src/transformers/__init__.py:839]: I will for now not include it in the init.

File: src/transformers/models/convnext/modeling_convnext.py (hunk 15 above)
Comment (NielsRogge) [reply src/transformers/models/convnext/modeling_convnext.py:405]: Hmm just checked, there is not a single vision model that has the `problem_type` defined.

Is it ok if I address that in a separate PR, for all vision models?

File: src/transformers/models/convnext/modeling_convnext.py [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+        # feature resolution stages, each consisting of multiple residual blocks
+        self.stages = nn.ModuleList()
+        dp_rates = [x.item() for x in torch.linspace(0, config.drop_path_rate, sum(config.depths))]
+        cur = 0
+        for i in range(config.num_stages):
+            stage = nn.Sequential(
+                *[
+                    ConvNextBlock(config, dim=config.dims[i], drop_path=dp_rates[cur + j])
+                    for j in range(config.depths[i])
+                ]
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/convnext/modeling_convnext.py:256]: maybe `stage` can be separated a `nn.Module`

**Review by FrancescoSaverioZuppichini — COMMENTED** [review]
Thanks for the addition, looks great! I've added a couple of comments

File: docs/source/model_doc/convnext.mdx [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+specific language governing permissions and limitations under the License.
+-->
+
+# ConvNeXT
+
+## Overview
+
+The ConvNeXT model was proposed in [A ConvNet for the 2020s](https://arxiv.org/abs/2201.03545) by Zhuang Liu, Hanzi Mao, Chao-Yuan Wu, Christoph Feichtenhofer, Trevor Darrell, Saining Xie.
+
+The abstract from the paper is the following:
```
Comment (LysandreJik) [inline docs/source/model_doc/convnext.mdx:19]: Could you also add a small blurb about what the model does/why it's special?

File: src/transformers/models/convnext/modeling_convnext.py [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+            Global average pooling of the last feature map followed by a layernorm.
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or
+        when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of each stage) of shape `(batch_size, num_channels,
+            height, width)`. Hidden-states of the model at the output of each layer.
+    """
+
+    last_hidden_state: torch.FloatTensor = None
+    pooler_output: Optional[torch.FloatTensor] = None
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
```
Comment (LysandreJik) [inline src/transformers/models/convnext/modeling_convnext.py:66]: This isn't added to the docs, in the model-specific outputs, and I'm not sure the format is correct

**Review by LysandreJik — COMMENTED** [review]
On top of the other comments, just two nits from my part

File: src/transformers/models/convnext/configuration_convnext.py [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    >>> configuration = model.config
+    ```"""
+    model_type = "convnext"
+
+    def __init__(
+        self,
+        num_channels=3,
+        num_stages=4,
+        dims=[96, 192, 384, 768],
+        depths=[3, 3, 9, 3],
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext/configuration_convnext.py:78]: In that case, we should also fix it for models like SegFormer.
